stingray/simulator/base.py

```diff
@@ -18,87 +28,152 @@ def simulate_times(lc, use_spline=False, bin_time=None):
+    ----------
+    time: array-like
+    counts: array-like
+    gti: [[gti00, gti01], ..., [gtin0, gtin1]]
+    dt: float
 
-        new_nev = len(random_ts)
-        ev_list[nev:nev + new_nev] = random_ts[:]
-        nev += new_nev
-        bin_start += max_bin
+    Other Parameters
+    ----------------
+    use_spline : bool
+        Approximate the light curve with a spline to avoid binning effects
 
-    # Discard all zero entries at the end
-    time = ev_list[:nev]
-    time.sort()
+    Returns
+    -------
+    times : array-like
+        Simulated photon arrival times
 
-    return time
+    Examples
+    --------
+    >>> t = [0.5, 1.5, 2.5, 3.5, 5.5]
+    >>> c = [100] * 5
+    >>> gti = [[0, 4], [5, 6]]
+    >>> times = simulate_times_from_count_array(t, c, gti, 1, use_spline=True)
+    >>> np.all(np.diff(times) > 0)  # Output array is sorted
+    True
+    >>> np.all(times >= 0.)  # All times inside GTIs
+    True
+    >>> np.all(times <= 6.)
+    True
+    >>> np.any(times > 5.)
+    True
+    >>> np.any(times < 4.)
+    True
+    >>> np.any((times > 4.) & (times < 5.))  # No times outside GTIs
+    False
+    >>> c[0] = -3.
+    >>> simulate_times_from_count_array(t, c, gti, 1)  # Test with one negative value in the lc
+    Traceback (most recent call last):
+        ...
+    ValueError: simulate_times can only work with...
+    """
+    time = np.asarray(time)
+    counts = np.asarray(counts)
+    gti = np.asarray(gti)
+    kind = "linear"
+    if use_spline and time.size > 2:
+        kind = "cubic"
+
+    if np.any(counts < 0):
+        raise ValueError(
+            "simulate_times can only work with positive-definite light curves"
+        )
+
+    if len(gti) > 1:  # Work GTI by GTI, to avoid the spillover of events
+        all_events = []
+        start_bins, stop_bins = gti_border_bins(gti, time, dt=dt)
+        for i, (start, stop) in enumerate(zip(start_bins, stop_bins)):
+            new_events = simulate_times_from_count_array(
+                time[start:stop],
+                counts[start:stop],
+                [gti[i]],
+                dt,
+                use_spline=use_spline)
+            all_events.append(new_events)
+        return np.concatenate(all_events)
+
+    if len(counts) == 1:  # Corner case: a single light curve bin
+        dt = dt
+        t0 = time[0] - dt / 2
+        t1 = time[0] + dt / 2
+        N = counts[0]
+        return np.sort(np.random.uniform(t0, t1, N))
+
+    tmin = gti[0, 0]
+    tmax = gti[-1, 1]
+    duration = (tmax - tmin)
+    phase_bins = np.copy(time)
+    phase_bins -= tmin
+    phase_bins /= duration
+    dph = dt / duration
+
+    counts = np.concatenate(([0], counts))
+    phase_bins = np.concatenate(
+        ([0], phase_bins + dph / 2))
+    n_events_predict = np.random.poisson(np.sum(counts))
+
+    cdf = np.cumsum(counts, dtype=float)
+    cdf -= cdf[0]
+    cdf /= cdf[-1]
+
+    inv_cdf_func = sci.interp1d(
+        cdf,
+        phase_bins,
+        kind=kind)
```

Comment:
Oh, I see, you need some kind of interpolation in order to get the inverse CDF to work. Are there noticeable changes if you use one versus the other?

Reply:
Not that I know of, but I figured this is a very likely thing to happen, so I added the possibility to interpolate smoothly.